pkg/dns/dns.go

```diff
@@ -25,54 +25,27 @@ type ServerOptions struct {
 	IPv6            bool
 	StaticHosts     map[string]string
 	UpstreamServers []string
-	Listener        net.Listener
 	TruncateReply   bool
 }
 
 type Handler struct {
 	truncate     bool
 	clientConfig *dns.ClientConfig
 	clients      []*dns.Client
-	IPv6         bool
-	cname        map[string]string
-	ip           map[string]net.IP
+	ipv6         bool
+	cnameToHost  map[string]string
+	hostToIP     map[string]net.IP
 }
 
 type Server struct {
 	udp *dns.Server
 	tcp *dns.Server
 }
 
-func Cname(host string) string {
-	host = strings.ToLower(host)
-	if !strings.HasSuffix(host, ".") {
-		host += "."
-	}
-	return host
-}
-
-func (s *Server) Shutdown() {
-	if s.udp != nil {
-		_ = s.udp.Shutdown()
-	}
-	if s.tcp != nil {
-		_ = s.tcp.Shutdown()
-	}
-}
-
-func newStaticClientConfig(ips []string) (*dns.ClientConfig, error) {
-	s := ``
-	for _, ip := range ips {
-		s += fmt.Sprintf("nameserver %s\n", ip)
-	}
-	r := strings.NewReader(s)
-	return dns.ClientConfigFromReader(r)
-}
-
-func newHandler(truncate, IPv6 bool, hosts map[string]string, upstreamServers []string) (dns.Handler, error) {
+func NewHandler(opts *ServerOptions) (dns.Handler, error) {
```

Comment:
> They are not really following the exported function/comment Scheme, I think it's best to leave it since it aligns with the rest of the project.

I think adding the comments will be fine, especially if it is in a part of the code that is re-used outside the project.